### PR TITLE
repo: pass and return ConnRef via pointers

### PR DESCRIPTION
**Pull request by stolowski**
Followup to #4358 (requested in the review) to unify ConnRef passing via pointers.
Should be reviewed after interface-hooks land.

### CODE REVIEW

**Comment by zyga** [comment]
@stolowski this needs de-conflicting

**Comment by bboozzoo** [comment]
Still conflicts.

File: interfaces/core.go [hunk 1]
```diff
@@ -133,22 +133,22 @@ func (conn *ConnRef) ID() string {
 }
 
 // ParseConnRef parses an ID string
-func ParseConnRef(id string) (ConnRef, error) {
+func ParseConnRef(id string) (*ConnRef, error) {
 	var conn ConnRef
 	parts := strings.SplitN(id, " ", 2)
 	if len(parts) != 2 {
-		return conn, fmt.Errorf("malformed connection identifier: %q", id)
```
Comment (zyga) [inline interfaces/core.go]: Nice touch

File: interfaces/repo.go [hunk 2]
```diff
@@ -691,42 +691,42 @@ func (r *Repository) connected(snapName, plugOrSlotName string) ([]ConnRef, erro
...
-func (r *Repository) Connections(snapName string) ([]ConnRef, error) {
+func (r *Repository) Connections(snapName string) ([]*ConnRef, error) {
 	r.m.Lock()
 	defer r.m.Unlock()
 
 	if snapName == "" {
 		snapName, _ = r.guessCoreSnapName()
 		if snapName == "" {
-			return nil, fmt.Errorf("internal error: cannot obtain core snap name while computing connections")
+			return nil, fmt.Errorf("snap name is empty")
```
Comment (zyga) [inline interfaces/repo.go]: Is this change warranted?

File: overlord/ifacestate/ifacestate.go [hunk 3]
```diff
@@ -195,6 +195,11 @@ func connect(st *state.State, change *state.Change, installedSnapTask *state.Tas
 	connectPlugConnection := hookstate.HookTask(st, summary, connectPlugHookSetup, initialContext)
 	connectPlugConnection.WaitFor(connectSlotConnection)
 
+	// store IDs of the connect-plug/slot- hook tasks in the main 'connect' task,
```
Comment (zyga) [inline overlord/ifacestate/ifacestate.go]: This looks unrelated to the PR

File: tests/main/interfaces-hooks/task.yaml [hunk 4]
```diff
@@ -1,5 +1,9 @@
 summary: Check that `snap connect` runs interface hooks
 
+environment:
```
Comment (zyga) [inline tests/main/interfaces-hooks/task.yaml]: This looks unrelated as well.

File: interfaces/repo_test.go [hunk 5]
```diff
@@ -1225,12 +1225,32 @@ func (s *RepositorySuite) TestConnectedFindsCoreSnap(c *C) {
...
 	c.Assert(err, IsNil)
 
 	conns, err := s.testRepo.Connected("", s.slot.Name)
 	c.Assert(err, IsNil)
-	c.Check(conns, DeepEquals, []ConnRef{*NewConnRef(s.plug, slot)})
+	c.Check(conns, DeepEquals, []*ConnRef{NewConnRef(s.plug, slot)})
+}
+
+// Connected finds connections when asked from plug or from slot side
+func (s *RepositorySuite) TestConnections(c *C) {
```
Comment (zyga) [inline interfaces/repo_test.go]: This looks ok, is it related to the change or just an addition from other PRs?

**Comment by zyga** [comment]
There are some incorrectly resolved conflicts here as well:
```
# github.com/snapcore/snapd/interfaces_test
interfaces/repo_test.go:1257:6: (*RepositorySuite).TestConnections redeclared in this block
	previous declaration at interfaces/repo_test.go:1237:6
interfaces/repo_test.go:1260:31: cannot use *interfaces.NewConnRef(s.plug, s.slot) (type interfaces.ConnRef) as type *interfaces.ConnRef in argument to s.testRepo.Connect
```

File: interfaces/repo.go (hunk 2 above)
Comment (stolowski) [reply interfaces/repo.go]: Nope, it's my error in merging, thanks for spotting!

File: interfaces/repo_test.go (hunk 5 above)
Comment (stolowski) [reply interfaces/repo_test.go]: Nah, it's merging mistake again, it's the function from below, only without all the changes I made. Thanks!

File: overlord/ifacestate/ifacestate.go (hunk 3 above)
Comment (stolowski) [reply overlord/ifacestate/ifacestate.go]: Indeed, thanks again for spotting, this is something I removed in parent PR (which already landed) and now erronously brought it back when de-conflicting.

File: tests/main/interfaces-hooks/task.yaml (hunk 4 above)
Comment (stolowski) [reply tests/main/interfaces-hooks/task.yaml]: Yep...

File: interfaces/repo.go [hunk 6]
```diff
@@ -439,27 +437,26 @@ func (r *Repository) ResolveConnect(plugSnapName, plugName, slotSnapName, slotNa
...
-		return ref, fmt.Errorf("snap %q has no slot named %q", slotSnapName, slotName)
+		return nil, fmt.Errorf("snap %q has no slot named %q", slotSnapName, slotName)
 	}
 	// Ensure that plug and slot are compatible
 	if slot.Interface != plug.Interface {
-		return ref, fmt.Errorf("cannot connect %s:%s (%q interface) to %s:%s (%q interface)",
+		return nil, fmt.Errorf("cannot connect %s:%s (%q interface) to %s:%s (%q interface)",
 			plugSnapName, plugName, plug.Interface, slotSnapName, slotName, slot.Interface)
 	}
-	ref = *NewConnRef(plug, slot)
```
Comment (bboozzoo) [inline interfaces/repo.go]: Nice
